src/ImageSharp/Formats/Tiff/Ifd/DirectoryReader.cs

```diff
@@ -14,92 +13,60 @@ namespace SixLabors.ImageSharp.Formats.Experimental.Tiff
     /// </summary>
     internal class DirectoryReader
     {
-        private readonly TiffStream stream;
+        private readonly ByteOrder byteOrder;
 
-        private readonly EntryReader tagReader;
+        private readonly Stream stream;
 
         private uint nextIfdOffset;
 
-        public DirectoryReader(TiffStream stream)
+        public DirectoryReader(ByteOrder byteOrder, Stream stream)
         {
+            this.byteOrder = byteOrder;
             this.stream = stream;
-            this.tagReader = new EntryReader(stream);
         }
 
-        public IEnumerable<IExifValue[]> Read()
+        public IEnumerable<ExifProfile> Read()
         {
-            if (this.ReadHeader())
-            {
-                return this.ReadIfds();
-            }
+            this.nextIfdOffset = new HeaderReader(this.byteOrder, this.stream).ReadFileHeader();
 
-            return null;
-        }
-
-        private bool ReadHeader()
-        {
-            ushort magic = this.stream.ReadUInt16();
-            if (magic != TiffConstants.HeaderMagicNumber)
-            {
-                TiffThrowHelper.ThrowInvalidHeader();
-            }
+            IEnumerable<List<IExifValue>> ifdList = this.ReadIfds();
 
-            uint firstIfdOffset = this.stream.ReadUInt32();
-            if (firstIfdOffset == 0)
+            var list = new List<ExifProfile>();
+            foreach (List<IExifValue> ifd in ifdList)
             {
-                TiffThrowHelper.ThrowInvalidHeader();
+                var profile = new ExifProfile();
+                profile.InitializeInternal(ifd);
+                list.Add(profile);
             }
 
-            this.nextIfdOffset = firstIfdOffset;
-
-            return true;
+            return list;
         }
 
-        private IEnumerable<IExifValue[]> ReadIfds()
+        private IEnumerable<List<IExifValue>> ReadIfds()
         {
-            var list = new List<IExifValue[]>();
+            var valuesList = new List<List<IExifValue>>();
+            var readersList = new SortedList<uint, EntryReader>();
             while (this.nextIfdOffset != 0)
             {
-                this.stream.Seek(this.nextIfdOffset);
-                IExifValue[] ifd = this.ReadIfd();
-                list.Add(ifd);
-            }
+                var reader = new EntryReader(this.byteOrder, this.stream, this.nextIfdOffset);
+                List<IExifValue> values = reader.ReadValues();
+                valuesList.Add(values);
 
-            this.tagReader.LoadExtendedData();
+                this.nextIfdOffset = reader.NextIfdOffset;
```

Comment:
we should check if `NextIfdOffset` is valid, e.g. not larger then the available data and throw an exception if it is. I have an example tiff image where this is the case: [Example](http://www.rootsweb.com/~irlcav/moy.tif). Otherwise the loop will not terminate here.

Reply:
Your test file seem damaged, I added correct handling of such, AsTiffTagViewer can read only the first frame too.

Here reading offset of the first frame:
caseshttps://github.com/SixLabors/ImageSharp/pull/1494/files/acf5f3b9646024522898255de1ec40cd38b89f0d#diff-81064e605436da75589b14b7e457c70b71490682ffdc6fd351c919484c7f14e1R30

to this line we have a readed offset:
https://github.com/SixLabors/ImageSharp/pull/1494/files/acf5f3b9646024522898255de1ec40cd38b89f0d#diff-81064e605436da75589b14b7e457c70b71490682ffdc6fd351c919484c7f14e1R49

Reply:
> Your test file seem damaged, I added correct handling of such, AsTiffTagViewer can read only the first frame too.

@IldarKhayrutdinov: yes it is quite possible that this file is indeede broken, but we should be able to handle it nonetheless (at least throw an exception indicating that it is broken).

Im not sure, if i understand you correctly: You said "I added correct handling of such", but i dont see any recent changes. Have you not pushed your changes yet, or do i miss something?

btw. this image is from a [TiffTestSuite](https://code.google.com/archive/p/imagetestsuite/source/default/source) which has various TIFF file which could be useful for testing.

Reply:
yes, I have not pushed yet, I am working on other remarks.

Reply:
> btw. this image is from a [TiffTestSuite](https://code.google.com/archive/p/imagetestsuite/source/default/source) which has various TIFF file which could be useful for testing.

a good set of test data, you need to add them to yourself

Reply:
@brianpopow I pushed, added your test file